### PR TITLE
backend sampling: support returning post-sampling probs

### PR DESCRIPTION
**Pull request by TimNN**
## Overview

Currently, if post-sampling probs are requested, backend sampling is forcibly disabled. 

This PR makes the two work together.

## Additional information

Similar to the added test case, I have verified that the returned probabilities are the same with and without backend sampling for some additional samplers and models and noted only the following differences:

* Small variations in probabilities, sometimes up to 1e-4 but often more in the 1e-6 to 1e-8 range.
* Different behaviors regarding whether 0.0 probabilities are included in the output.
  * For example, the Min P sampler includes 0.0 probabilities only for backend sampling, but 0.0 temperature sampling includes 0.0 probabilities only without backend sampling.
* (To get these results, a minor change (not included in this PR)  was needed to make the dist sampler implementations consistent and always generate a random number, even if there was only one element). 

I'm not confident that the added test case is the best way to test this. If you have suggestions for a better test, please let me know.

I also tried to make logprobs (i.e. `post_sampling_probs: false`) work, but couldn't figure out how to get the same result as without backend sampling. AFAICT, the logits returned by the API without backend sampling (`llama_get_logits_ith`) are the pre-sampling logits, but the backend sampling API (`llama_get_sampled_logits_ith`) always returns the post-sampling logits.

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: NO (except maybe some CLion autocomplete)

### CODE REVIEW

**Comment by ggerganov** [comment]
> I also tried to make logprobs (i.e. post_sampling_probs: false) work, but couldn't figure out how to get the same result as without backend sampling. AFAICT, the logits returned by the API without backend sampling (llama_get_logits_ith) are the pre-sampling logits, but the backend sampling API (llama_get_sampled_logits_ith) always returns the post-sampling logits.

Currently, we don't extract the pre-sampling logits when all sequences use backend sampling:

https://github.com/ggml-org/llama.cpp/blob/29debb3a6a4c291d66aabbc46a0bb8c17a77e267/src/llama-context.cpp#L1737-L1750

See the `needs_raw_logits()` logic.

One of the main reason to use backend sampling is to reduce the device-to-host data transfers, so therefore the pre-sampling logits are not extracted.

**Comment by TimNN** [comment]
Ah, that makes sense.

---

I saw the test failures:

* `unit/test_completion.py::test_n_probs_post_sampling` passes locally. The assertion failure looks to me like the test is now running with backend_sampling enabled (which for the samplers used by the test means that 0.0 probabilities are no longer returned). I may have misunderstood something about how the global state in the Python test is handled. I'll investigate some more.
* `test-llama-archs`: At a first glance, this test doesn't depend on any of the code I'm modify, so my current assumption is that this is a false-positive / the test is flaky, but I'll try to reproduce.

**Comment by ggerganov** [comment]
In this workflow, the backend sampling is enabled for all tests through the global environment variable:

https://github.com/ggml-org/llama.cpp/actions/runs/25627363312/workflow?pr=22622#L70

The goal is to exercise all the server functionality via the e2e tests when the backend sampling is enabled. Having a specific test for backend sampling similar to the one you added is fine. We just need to figure out how to make the rest of the tests also be compatible with the global setting.

> The assertion failure looks to me like the test is now running with backend_sampling enabled (which for the samplers used by the test means that 0.0 probabilities are no longer returned). I may have misunderstood something about how the global state in the Python test is handled. I'll investigate some more.

Yes, that's most likely it. How should we address this?
I have some very vague memory about this issue when we were developing the backend sampling, but can't remember anything specific atm. Most likely we want the sampling to behave similar in both regimes, but I think there was something non-trivial preventing this.

**Comment by TimNN** [comment]
> Yes, that's most likely it. How should we address this?

To ensure a consistent server response, we could just not include 0.0 probabilities in the server response, e.g.: https://github.com/TimNN/llama.cpp/commit/a76d335c345e21ee80fb41eaf22cafefa762bfa6

0.0 probability tokens are IMO somewhat meaningless anyway since which ones you actually get can be somewhat random. The server README documentation allows for that already, I believe:

> Number of elements in top_probs may be less than n_probs

If that seems reasonable to you, I can prepare a separate PR for that and we can merge that one first.

---

The `test_n_probs_post_sampling` test currently expects exactly `n_probs = 10` probabilities to be returned despite opting in to greedy sampling with `temperature = 0.0` (which IIUC means that there's always going to be exactly 1 token with probability 1.0 and all others must be probability 0.0). 

I'm not sure of the intent behind that. 

If the test wants to verify real responses with multiple tokens, IMO it should run with a higher temperature.

**Comment by ggerganov** [comment]
> If that seems reasonable to you, I can prepare a separate PR for that and we can merge that one first.

OK, sounds good. Can add the change directly in this PR.

> The test_n_probs_post_sampling test currently expects exactly n_probs = 10 probabilities to be returned despite opting in to greedy sampling with temperature = 0.0 (which IIUC means that there's always going to be exactly 1 token with probability 1.0 and all others must be probability 0.0).
>
> I'm not sure of the intent behind that.
>
> If the test wants to verify real responses with multiple tokens, IMO it should run with a higher temperature.

The idea is probably to get some deterministic probability distribution. But since the tests run on variety of hardware, there isn't a good way to do it and hence a basic `temp = 0` sampling is used. If you have any ideas to improve this - feel free to propose.

**Comment by TimNN** [comment]
> OK, sounds good. Can add the change directly in this PR.

Done.

> If you have any ideas to improve this - feel free to propose.

I've updated the test to run at a higher temperature and be a bit more flexible in what it accepts. (I patched the test locally to run with 1000 different seeds to simulate different results on different hardware and it still passes reliably).

The server unit tests pass locally with and without `LLAMA_ARG_BACKEND_SAMPLING=1`.

**Review by ggerganov — APPROVED** [review]

**Review by ServeurpersoCom — APPROVED** [review]
